File: 백준/Silver/2003. 수들의 합 2/수들의 합 2.py
from sys import stdin

n, m = map(int, stdin.readline().split())
numbers = list(map(int, stdin.readline().split()))
start, end, number_sum = 0, 0, numbers[0]
number_len = len(numbers)
count = 0

while True:
    if number_sum == m:
        count, end = count + 1, end + 1
        if end >= number_len:
            break
        number_sum += numbers[end]
    elif number_sum < m:
        end += 1
        if end >= number_len:
            break
        number_sum += numbers[end]
    else:
        number_sum, start = number_sum - numbers[start], start + 1
        if start >= number_len:
            break
    
print(count)

File: 백준/Silver/2003. 수들의 합 2/수들의 합 2.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken()), m = Integer.parseInt(st.nextToken());
        int nums[] = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0 ; st.hasMoreTokens(); i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }
        int start = 0, end = 0, count = 0;
        while (end < n && start < n) {
            for (; end < n && m - nums[end] >= 0 ; end++) {
                m -= nums[end];
            }
            if (m == 0) {
                count++;
            }
            m += nums[start++];
        }
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        bw.write(String.valueOf(count));
        bw.flush();
    }
}